text/0009-multilink_command_handling.md

```diff
@@ -0,0 +1,36 @@
+  * Start date: 2018-05-21
+  * Contributors: Achermann Florian, <acfloria@ethz.ch>
+  * Related issues: 
+  
+# Summary
+
+The goal of this document is to present a design on how to handle commands which are sent over different links with significantly different latencies from the ground control station to a vehicle.
+  
+# Motivation
+
+When transmitting commands over multiple links with significantly different latencies to a single vehicle the order in which the commands are sent might not be equal to the order in which they are received. This can lead to an undesired response of the vehicle especially when using commands such as MAV_CMD_DO_CHANGE_ALTITUDE, MAV_CMD_DO_SET_MODE, or MAV_CMD_DO_REPOSITION.
+
+# Detailed Design
+
+A new type of command is created: COMMAND_STAMPED. This command is basically a COMMAND_LONG with additional metadata:
+  * (uint64) vehicle_time: The vehicle time since boot in ms
+  * (uint64) utc_time: The seconds elapsed since Jan 01 1970 (UTC)
+
+The vehicle_time field is filled in with the value of the timestamp field of the last received vehicle time on the groundstation (SYSTEM_TIME or HIGH_LATENCY2). This indicates on which vehicle status the command is sent. On the groundstation the last vehicle is reset to 0 after some time not receiving an update to avoid that any sent command is rejected.
+
+The utc_time provides a reference between different flights. Which is necessary as for example Satellite Communication messages can be queued on the network and arrive on the plane on the next flight.
```

Comment:
Unless `vehicle_time == 0` is a special case then you can't queue messages in this manner, as they won't have a `vehicle_time` tag that fits within the valid time slot window when the vehicle gets to actually parsing the message.

Reply:
I do not really get what you mean here.

Yes, `vehicle_time == 0` is a special case. In that case the commands are not queued only the checks regarding the `utc_time` field are executed. It is only the case if no `SYSTEM_TIME` or `HIGH_LATENCY2` has yet been received on the ground station. This case is added that this station still can send commands which are executed.

Reply:
The problem with that is you specified that a GCS should reset `vehicle_time` to 0 if it hasn't seen a packet in awhile. That means a high latency link (with a lot of variation in delivery time) could be repeatedly setting the vehicle_time to 0, because it encountered a time out, in which case you are by design giving up on the multilink command handling.

Reply:
True, I thought of a timeout which is quite large. But by default I would anyway disable resetting the `vehicle_time` but I think such a mechanism could be useful.